cloudant-sync-datastore-android/src/main/java/com/cloudant/sync/replication/PeriodicReplicationService.java

```diff
@@ -417,8 +444,51 @@ private int getIntervalInSeconds() {
         }
     }
 
-    private String constructKey(String suffix) {
-        return getClass().getName() + suffix;
+    String constructKey(String suffix) {
+        return constructKey(getClass(), suffix);
+    }
+
+    static String constructKey(Class<? extends PeriodicReplicationService> prsClass,
+                                       String suffix) {
+        return prsClass.getName() + suffix;
+    }
+
+    @Override
+    public void allReplicationsCompleted() {
+        super.allReplicationsCompleted();
+        setReplicationsPending(this, getClass(), false);
+    }
+
+    /**
+     * Sets whether there are replications pending. This may be because replications are
+     * currently in progress and have not yet completed, or because a previous scheduled
+     * replication didn't take place because the conditions for replication were not met.
+     * @param context
+     * @param prsClass
+     * @param pending true if there is a replication pending, or false otherwise.
+     */
+    public static void setReplicationsPending(Context context, Class<? extends
```

Comment:
Your understanding is correct, and indeed, there was a small bug (fixed in 48f5b55) that meant pending replications wouldn't always be resumed immediately we reconnected to WiFi following a reboot because the persistent `wasOnWifi` flag stored in shared preferences wasn't being reset following reboot. Under this circumstance, replications sometimes wouldn't resume until the next scheduled period. However, I think under most circumstances when the device turned off, we would probably have got an event indicating we were disconnecting from WiFi as the device shutdown, so I think it's likely this bug would only affect a device crash or forced shutdown, and everything would have got back to normal following the next scheduled replication.

Thanks for raising this issue and making me go back to look at it.

I've fixed this by setting the `wasOnWifi` flag to `false` when we get notification of the reboot. This will be safe as we'll always get the notification of the reboot before we get notification that the device has connected to WiFi following the reboot.

Now, with this fix, if the device reboots when there are still replications pending:

1. `WifiPeriodicReplicationReceiver` will receive an Intent with action `Intent.ACTION_BOOT_COMPLETED` and we'll make sure the previous state of the WiFi is set so that it shows we weren't on WiFi.
1. Then WiFi will either be available or not available:
    * If WiFi is available when the device restarts, `WifiPeriodicReplicationReceiver` will then receive an Intent with action `ConnectivityManager.CONNECTIVITY_ACTION`. It will then go into the `else if (isConnectedToWifi)` branch and check `replicationsPending` and immediately resume the replication.
    * If WiFi was not available when the device restarted, when we next connect to WiFi, the same will happen. If there are any `ALARM_ACTION` intents received before the `ConnectivityManager.CONNECTIVITY_ACTION` indicating we're back on WiFi, the `replicationsPending` will be re-set to true (although it will already be true anyway).
1. When the replications complete, the `allReplicationsCompleted()` callback in `PeriodicReplicationService` will set `replicationsPending` to `false`.